source/shared/cpp/ObjectModel/MarkDownParser.cpp

```diff
@@ -0,0 +1,425 @@
+                if (stk.empty() && top_left != m_leftLookUpTable.end())
+                {
+                    stk.push_back(top_left);
+                    top_left++;
+                }
+            }
+        }
+    }
+
+    // append all processed tags
+    std::ostringstream html;
+    auto leftItr = m_leftLookUpTable.begin(), rightItr = m_rightLookUpTable.begin();
+    for (auto idx = 0; idx < m_tokenizedString.size (); idx++)
+    { 
+        if (leftItr != m_leftLookUpTable.end() && leftItr->m_idx == idx)
+        {
+            // if there are unused emphasis, append them 
+            if (leftItr->m_emphCnts)
+            {
+                int startIdx = m_tokenizedString[idx].size () - leftItr->m_emphCnts;
+                html << m_tokenizedString[idx].substr (startIdx, std::string::npos);
+            }
+            // append tags; since left delims, append it in the reverse order
+            for (auto itr = leftItr->m_tags.rbegin(); itr != leftItr->m_tags.rend(); itr++)
+            { 
+                html << ((*itr == EmphasisItalic)? "<em>" : "<strong>");
+            }
+            leftItr++;
+        } 
+        else if (rightItr != m_rightLookUpTable.end() && rightItr->m_idx == idx)
+        {
+            // append tags; 
+            for (auto itr = rightItr->m_tags.begin(); itr != rightItr->m_tags.end(); itr++)
+            { 
+                html << ((*itr == EmphasisItalic)? "</em>" : "</strong>");
+            }
+            // if there are unused emphasis, append them 
+            if (rightItr->m_emphCnts)
+            {
+                int startIdx = m_tokenizedString[idx].size () - rightItr->m_emphCnts;
+                html << m_tokenizedString[idx].substr (startIdx, std::string::npos);
+            }
+            rightItr++;
+        }
+        else
+        {
+            html << m_tokenizedString[idx];
+        }
+    }
+
+    return "<p>" + html.str () + "</p>";
+}
+
+void MarkDownParser::GetCh(char ch)
+{
+    // store chars unless it's escape
+    if (ch != '\\')
+    {
+        m_tokenizedString[m_currentWordIndex] += ch;
+    }
+}
+
+void MarkDownParser::PutBackCh()
+{
+    m_tokenizedString.pop_back();
+    --m_currentWordIndex;
+    char ch = (m_currentDelimiterType == Asterisk)? '*' : '_';
+    // last ch gets stored; so no need to put back last ch
+    if (m_curPos + 1 == m_textEnd)
+    { 
+        m_delimiterCnts--;
+    }
+
+    while (m_delimiterCnts--)
+    {
+        GetCh(ch);
+    }
+}
+
+bool MarkDownParser::IsLeftEmphasisDelimiter () 
+{
+    bool isLeftEmphasisDetecting = ((m_EmphasisState & ~EmphasisOn) && 
+                          m_delimiterCnts && 
+                          !isspace (*m_curPos) && 
+                          !(m_lookBehind == Alphanumeric && ispunct(*m_curPos)) && 
+                          !(m_lookBehind == Alphanumeric && m_currentDelimiterType == Underscore));
+    if (isLeftEmphasisDetecting)
+    {
+        m_leftLookUpTable.push_back(
```

Comment:
>m_leftLookUpTable [](start = 8, length = 17)

Functions called Is* that return a bool shouldn't have side effects.  Lets either rename this function to describe what it's actually doing (and make isLeftEmphasisDelimiter an out parameter) or move the part that's updating the member variables someplace else.